src/multivariate/mvdiscretenonparametric.jl

```diff
@@ -0,0 +1,153 @@
+struct MvDiscreteNonParametric{T <: Real,P <: Real,Ts <:  ArrayOfSimilarArrays{T},Ps <: AbstractVector{P}} <: DiscreteMultivariateDistribution
```

Comment:
Can we make this more general and change it to something like
```suggestion
struct GeneralDiscreteNonParametric{VF,T,P<:Real,Ts<:AbstractVector{T},Ps<:AbstractVector{P}} <: Distribution{VF,Discrete}
```
? Then it could also be used instead of `DiscreteNonParametric` for the univariate case and we could define the alias
```julia
const DiscreteNonParametric{T<:Real,P<:Real,Ts<:AbstractVector{T},Ps<:AbstractVector{P}} = GeneralDiscreteNonParametric{Univariate,T,P,Ts,Ps}
```
to make it a non-breaking change.

Reply:
And one could also define the alias
```suggestion
const MvDiscreteNonParametric{T<:AbstractVector{<:Real},P<:Real,Ts<:AbstractVector{T},Ps<:AbstractVector{P}} = GeneralDiscreteNonParametric{Multivariate,T,P,Ts,Ps}
```

Reply:
Can you expand on how the `GeneralDiscreteNonParametric` shoudl work? It's not clear to me how this `VF` work, for example, I tried doing the following:
```julia

struct GeneralDiscreteNonParametric{VF,T,P<:Real,Ts<:AbstractVector{T},Ps<:AbstractVector{P}} <: Distribution{VF,Discrete}

    support::Ts
    p::Ps

    function GeneralDiscreteNonParametric{VF,T,P,Ts,Ps}(support::Ts,
        p::Ps) where {T,P <: Real,Ts <: AbstractVector{<:AbstractVector{T}},Ps <: AbstractVector{P}}

        length(support) == length(p) || error("length of `support` and `p` must be equal")
        isprobvec(p) || error("`p` must be a probability vector")
        allunique(support) || error("`support` must contain only unique value")
        new{VF,T,P,Ts,Ps}(support, p)
    end
end
```
But of course I get an error, since `VF` is not defined.

Reply:
Also, should the `GeneralDiscreteNonParametric` dispatch to `MvDiscreteNonParametric` and `DiscreteNonParametric`? Would something like CategoricalDistributions.jl be inside `GeneralDiscreteNonParametric`?

Reply:
@devmotion , should I leave the functions for `MvDiscreteNonParametric` or should I drop it and just do everything inside the `GeneralMvDiscreteNonParametric`?

Reply:
> But of course I get an error, since VF is not defined.

You have to provide it, like
```julia
struct GeneralDiscreteNonParametric{VF,T,P<:Real,Ts<:AbstractVector{T},Ps<:AbstractVector{P}} <: Distribution{VF,Discrete}
    support::Ts
    p::Ps

    function GeneralDiscreteNonParametric{VF,T,P,Ts,Ps}(support::Ts,
        p::Ps; check_args=true) where {VF,T,P<: Real,Ts<:AbstractVector{T},Ps<:AbstractVector{P}}
        if check_args
            length(support) == length(p) || error("length of `support` and `p` must be equal")
            isprobvec(p) || error("`p` must be a probability vector")
            allunique(support) || error("`support` must contain only unique values")
        end
        new{VF,T,P,Ts,Ps}(support, p)
    end
end
```
Then you could define a univariate discrete non-parametric distribution with `GeneralDiscreteNonParametric{Univariate,Float64,Float64,Vector{Float64},Vector{Float64}}([-10.0, -3.4, 2.1, 3.5], [0.4, 0.2, 0.1, 0.1])`. But, of course, users shouldn't have to do this. Therefore we could use the aliases (also for the multivariate case) and define outer constructors such as
```julia
function DiscreteNonParametric(support::AbstractVector{<:Real}, p::AbstractVector{<:Real}; issorted=false)
    _p = issorted ? p : sort(p)
    return DiscreteNonParametric{eltype(support),eltype(_p),typeof(support),typeof(_p)}(support, _p)
end
```

Reply:
Besides Multivariate and Univariate, what kind of distribution you intend to keep under GeneralDiscreteNonParametric? Categorical Distributions?

Reply:
They are already a type alias of `DiscreteNonParametric`, so if `DiscreteNonParametric` works they should just work automatically.